core/domain/caching_services_test.py

```diff
@@ -0,0 +1,172 @@
+
+    def test_flush_cache_wipes_cache_clean(self):
+        """Tests whether flushing the cache removes the elements in the
+        cache.
+        """
+        key_value_mapping = {'a': '1', 'b': '2', 'c': '3'}
+        caching_services.set_multi(
+            caching_services.CACHE_NAMESPACE_DEFAULT, None, key_value_mapping)
+        exploration_key = 'key'
+        default_exploration = (
+            exp_domain.Exploration.create_default_exploration(
+                'exp_id_1', title='A title', category='A category'))
+        caching_services.set_multi(
+            caching_services.CACHE_NAMESPACE_EXPLORATION,
+            python_utils.convert_to_bytes(0),
+            {
+                exploration_key: default_exploration
+            })
+        caching_services.flush_memory_cache()
+        self.assertEqual(
+            caching_services.get_multi(
+                caching_services.CACHE_NAMESPACE_DEFAULT, None,
+                ['a', 'b', 'c']), {})
+        self.assertEqual(
+            caching_services.get_multi(
+                caching_services.CACHE_NAMESPACE_EXPLORATION,
+                python_utils.convert_to_bytes(0),
+                [exploration_key]),
+            {})
+
+    def test_invalid_namespace_raises_error(self):
+        invalid_namespace = 'invalid'
+        key_value_mapping = {'a': '1', 'b': '2', 'c': '3'}
+
+        with self.assertRaisesRegexp(
+            ValueError,
+            'Invalid namespace: %s.' % invalid_namespace):
+            caching_services.set_multi(
+                invalid_namespace, None,
+                key_value_mapping)
+
+        with self.assertRaisesRegexp(
+            ValueError,
+            'Invalid namespace: %s.' % invalid_namespace):
+            caching_services.get_multi(
+                invalid_namespace, None,
+                ['a', 'b', 'c'])
+
+        with self.assertRaisesRegexp(
+            ValueError,
+            'Invalid namespace: %s.' % invalid_namespace):
+            caching_services.delete_multi(
+                invalid_namespace, None, ['a', 'b', 'c'])
+
+    def test_get_multi_correctly_retrieves_cache_elements(self):
+        key_value_mapping = {'a': '1', 'b': '2', 'c': '3'}
+        caching_services.set_multi(
+            caching_services.CACHE_NAMESPACE_DEFAULT, None, key_value_mapping)
+        exploration_key = 'key'
+        default_exploration = (
+            exp_domain.Exploration.create_default_exploration(
+                'exp_id_1', title='A title', category='A category'))
+        caching_services.set_multi(
+            caching_services.CACHE_NAMESPACE_EXPLORATION,
+            python_utils.convert_to_bytes(0),
+            {
+                exploration_key: default_exploration
+            })
+        result = caching_services.get_multi(
+            caching_services.CACHE_NAMESPACE_DEFAULT, None, ['a', 'b', 'c'])
+        self.assertEqual(result, key_value_mapping)
+        result = caching_services.get_multi(
```

Comment:
Tests are a bit weak. What about partial fetches -- some valid and some invalid? What about fetches for the wrong namespace?

Probably should split into more separate test functions for particular cases -- "correctly retrieves cache elements" is not very specific.

Both comments apply for the functions below.

Reply:
Done.